hpx/lcos/local/futures_factory.hpp

```diff
@@ -265,7 +265,7 @@ namespace hpx { namespace lcos { namespace local {
                     parallel::execution::post(*exec_,
                         util::deferred_call(
                             &base_type::run_impl, std::move(this_)),
-                        schedulehint, annotation);
+                        exec_->get_schedulehint(), annotation);
```

Comment:
Why does the `schedulehint` provided by the executor takes precedence over the one passed in from the caller? Do we have this behavior everywhere?

Reply:
The one from the caller is always just a default param. The whole thing is a mess.

Reply:
Why making it an even bigger mess, then? Isn't there a better way?

Reply:
I don't believe this makes it a bigger mess. This fixes a bug where scheduling hints were not being used when set via the executor. In my opinion the mess is caused by a very complex system of overloads on executors for different methods of invocations and default parameters used in many places which hides invocations that might be clumsy or wasteful. Ideally the default params should be removed and the functions invoked with the right parameters - or an invocation signature which does not use the parameters should exist - unfortunately, invocations with hints, without hints, with annotation, without them, with policies and without them etc make the whole thing messy. I'd like to see a more consistent way of passing policies, parameters, scheduling information/hints and any other required parsm to be designed. I hope that as we move over to the new executor proposal we can clean up some of the existing code - but that's too much for this simple fix and deserves a bigger PR at some later date.

Reply:
Could you please be a bit more specific about what functions should not have default arguments for their parameters?

Reply:
I cannot provide such a list. I am simply trying to fix bugs that prevent me from optimizing and debugging my code. Many of the default params for `schedulehint` were introduced by me, and I take full blame for them - they were introduced because there is a lot of code to understand at the executor dispatch level that is not commented or easily understandable by anyone other than the person who wrote it - it was easier to get the functionality I need by adding these defaults than to edit the many places that would need to be changed.

I hope that the executor cleanups and new proposals make this simpler to maintain.

Reply:
I think long term futures factory should not have the option to take an executor (this functionality seems to be there mostly for the old thread executors). If one wants to create a task with an executor there is one way to do so: through `post` or `async_execute`. The handling of the hints will be more direct after #4301 (I'm also working on unifying the various launch parameters into a struct).